Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,23 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+def truncate_text_to_approximate_size(s, max_bytes, opts = {})
+  front_weight = opts[:front_weight] || 0.5
+  glue = opts[:gloe] || "\n[...snip...]\n"
+  return s if s.bytesize <= max_bytes
+  if front_weight < 0 || front_weight > 1
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  scale = 0.99 # just a tad small to allow for the glue
+  while true
+    front = s[0..((scale * max_bytes * front_weight).floor)]
+    back_start = s.size-((scale * max_bytes * (1.0-front_weight)).ceil)
+    back = s[back_start..s.size]
+    out = front + glue + back
+    return out if out.bytesize <= max_bytes
```

Comment:
I don't see a way to do this cleanly. I could move the "raise ... if" up to the top of the loop, leaving the `scale = scale - 0.05` at the bottom. But that means the return is still not the last statement. And if I move the `scale = ...` statement elsewhere, that means that either a) at the point of exit, the `scale` value is inconsistent with the rest of the variables, or b) we start from `scale = 1.04` outside the loop and decrement at the top, which is IMHO both ugly and less readable, because the `scale` variable is initialized to a value which it should never take on in use, for the sake of control flow.

An early exit from an open-ended loop kind of makes sense to me. If it reads better, maybe we could constrain the loop and pull the `raise` outside of it, so at least the `return` is the last statement inside the loop body.